Implement Cache and __cuda_array_interface__ for v21.06.00

gigony opened the pull request:
I'm sorry for this big change.
This includes CI build failure fixes as well as new features/fixes/documents that couldn't be merged due to CI build failures.

@quasiben on top of your change (https://github.com/rapidsai/cucim/pull/46), the only missing piece was, adding `pytest` (https://github.com/rapidsai/cucim/pull/48/commits/8334f1f2d46738c31108ebc4303812ab9d93da0c) (which was not revealed until I added `set -x` to the build script (https://github.com/rapidsai/cucim/pull/48/commits/fae529d1589b5db98f55cc3fb941552d25d76ff9).

@jakirkham This change doesn't include a doc builder change in your PR (
https://github.com/rapidsai/cucim/pull/45/files#diff-5ed0e3039eb400d13dc55579a034925ea8d08e9f2054085b1c79fdc46a7c09a8L29) so the script may need to further updated in the next change.

- Fix memory leak in deflate decoder (#37)
- Update build script to handle failures (#45, #46)
- Fix typo in meta.yaml (#45, #46)
- Add pytest to gpu build script (#45, #46)
- Implement cache mechanism (#34)
- Add `__cuda_array_interface__`
- Support CUDA 10.x for cucim.skimage package (#44)
- Expose preferred_memory_capacity() for cache
- Fix failures due to GIL and nullptr access
- Add notebooks for cache and cuda array interface (#34)
- Update change logs for v21.06.00

<!--

Thank you for contributing to ___PROJECT___ :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on main/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against main they should be resolved by merging main
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

gigony commented:
Hi @ajschmidt8, could this change be merged into 21.06.00 branch before release?

quasiben commented on python/cucim/setup.py:
```diff
@@ -49,14 +44,13 @@ def read(*names, **kwargs):
     packages=find_packages('src'),
     package_dir={'cucim': 'src/cucim'},
-    py_modules=[splitext(basename(path))[0] for path in glob('src/*.py')],
+    # # https://docs.python.org/3/distutils/setupscript.html
```
Any reason to leave these comments here instead of removing ?  

Same question for the `package_data=` commented code below

gigony replied:
Oh, no specific reason. Just for future reference. Will remove it. Thank you for pointing out that!

quasiben commented:
I found it easier to look at the individual commits.  I left a few small comments on the `setup.py` file.  I think there is probably better to do than add the `pytest`  conda package but i'm ok pushing that off for another PR

jakirkham commented on cpp/src/io/device_type.cpp:
```diff
@@ -0,0 +1,106 @@
+// https://en.cppreference.com/w/cpp/algorithm/find
+#if __cplusplus < 202002L
+template <class InputIt, class UnaryPredicate>
+constexpr InputIt myfind_if(InputIt first, InputIt last, UnaryPredicate p)
```
Seems like we had this earlier. Should we refactor this code if it is needed? Or if `constexpr` is not needed, we can use from STL. If we do need `constexpr`, maybe Boost also has this and we can use from there?

gigony replied:
Refactored out as `cucim/cpp/include/cucim/cpp20/find_if.h` file. and reuse the header now.
Thank you for pointing out!

jakirkham commented on cpp/src/io/format/image_format.cpp:
```diff
@@ -162,6 +162,13 @@ ImageMetadata& ImageMetadata::level_downsamples(const std::pmr::vector<float>& l
 
+ImageMetadata& ImageMetadata::level_tile_sizes(const std::pmr::vector<uint32_t>& level_tile_sizes)
+{
+    level_tile_sizes_ = std::move(level_tile_sizes);
```
Can we use `std::move` on `const &`? Or should the input be `&&`?

gigony replied:
Thanks a lot @jakirkham for pointing out that !!!
Without your review, it would end up with inefficient code (copy instead of move).

Thank you so much!

jakirkham commented on cpp/src/memory/memory_manager.cu:
```diff
@@ -74,4 +68,53 @@ void get_pointer_attributes(PointerAttributes& attr, const void* ptr)
     }
 }
 
+bool move_raster_from_host(void** target, size_t size, cucim::io::Device& dst_device)
```
Can device be `const`?

```suggestion
bool move_raster_from_host(void** target, size_t size, const cucim::io::Device& dst_device)
```

gigony replied:
Yes, it should be `const`. Will update it.

jakirkham commented on cpp/src/memory/memory_manager.cu:
```diff
@@ -74,4 +68,53 @@ void get_pointer_attributes(PointerAttributes& attr, const void* ptr)
+    return true;
+}
+
+bool move_raster_from_device(void** target, size_t size, cucim::io::Device& dst_device)
```
```suggestion
bool move_raster_from_device(void** target, size_t size, const cucim::io::Device& dst_device)
```

jakirkham commented on python/cucim/setup.py:
```diff
@@ -31,13 +28,11 @@ def read(*names, **kwargs):
 opts = dict(
     name='cucim',
     version=read('VERSION').strip(),  # versioneer.get_version(),
-    # cmdclass=versioneer.get_cmdclass()
+    cmdclass=versioneer.get_cmdclass(),
```
If we are using versioneer for `cmdclass`, should we use for `version` as well?

gigony replied:
In case of versioneer, version comes from git tags, but C++ code's build system currently requires version as `VERSION` file.
I can try to refactor C++ code builder to get the version from git tags later (https://ipenguin.ws/2012/11/cmake-automatically-use-git-tags-as.html) but the current VERSION file seems okay to me.

Currently, version number from versioneer is used only when CuPy with scikit-image is available (https://github.com/rapidsai/cucim/issues/44, https://github.com/rapidsai/cucim/pull/48/commits/818dc2f00416aa7ec07aabc27c460948661ce510#diff-80a85bd6c52ce89c0dc59a606ca1205a2673c20ff860c65a3964db51ffa64ec8R47)

jakirkham commented on python/cucim/src/cucim/__init__.py:
```diff
@@ -31,4 +31,19 @@
 """
 
-from .clara import CuImage, __version__, cli
+# Try to import cupy first.
+# If cucim.clara package is imported first, you may see the following error when running on CUDA 10.x (#44)
+#   python3: Relink `/usr/lib/x86_64-linux-gnu/libnccl.so.2.8.3' with `/lib/x86_64-linux-gnu/librt.so.1' for IFUNC symbol `clock_gettime'
+#   Segmentation fault
+try:
+    import cupy
+except ImportError:
+    pass
```
Interesting. Do we know why that is?

gigony replied:
Don't know the exact root cause, but I guess,

This change is related to https://github.com/rapidsai/cucim/issues/44

If the system has only CUDA 10.x library, since cucim.clara part of the library depends on CUDA 11 and different version of runtime, it looks that it can conflict with symbols CUDA-based library that depends on CUDA 10.X

You can see version-related issues from other Github issue:
https://github.com/tensorflow/tensorflow/issues/19375

jakirkham commented on python/cucim/src/localtest.py:
```diff
@@ -109,34 +110,34 @@ def load_tile_cucim(slide, start_loc, tile_size):
 
 openslide_tot_time = 0
 cucim_tot_time = 0
-for num_workers in range(1, num_threads + 1):
+for num_workers in (num_threads,):  # range(1, num_threads + 1):
```
Why is this commented?

gigony replied:
It is modified during the local test (since I didn't want to execute it for all available thread # configurations.)
Reverted it.

jakirkham reviewed (commented):
Thanks Gigon! 😄

Had some questions below

gigony commented on python/cucim/src/localtest.py:
```diff
@@ -109,34 +110,34 @@ def load_tile_cucim(slide, start_loc, tile_size):
-                )
-            openslide_time = timer.elapsed_time()
-            openslide_tot_time += openslide_time
+    #     count = 0
+    #     for h in range(start_location, height, tile_size):
+    #         for w in range(start_location, width, tile_size):
+    #             count += 1
+    #     start_loc_iter = ((w, h)
+    #                       for h in range(start_location, height, tile_size)
+    #                       for w in range(start_location, width, tile_size))
+    #     with Timer("  Thread elapsed time (OpenSlide)") as timer:
+    #         with concurrent.futures.ThreadPoolExecutor(
+    #             max_workers=num_workers
+    #         ) as executor:
+    #             executor.map(
+    #                 lambda start_loc: load_tile_openslide(
+    #                     slide, start_loc, tile_size),
+    #                 start_loc_iter,
+    #             )
+    #         openslide_time = timer.elapsed_time()
+    #         openslide_tot_time += openslide_time
```
my bad. changed the code during the local testing. Will revert it. Thanks!

gigony commented on python/cucim/src/localtest.py:
```diff
@@ -147,8 +148,8 @@ def load_tile_cucim(slide, start_loc, tile_size):
         cucim_tot_time += cucim_time
-    print("  Performance gain (OpenSlide/cuCIM): {}".format(
-        openslide_time / cucim_time))
+    # print("  Performance gain (OpenSlide/cuCIM): {}".format(
+    #     openslide_time / cucim_time))
```
Will revert it.

gigony commented:
Thanks @jakirkham and @quasiben for the review! I would address your comments tonight.

gigony commented on python/cucim/setup.cfg:
```diff
@@ -2,7 +2,7 @@
 VCS = git
 style = pep440
 versionfile_source = src/cucim/_version.py
```
I don't think so.
`_version.py` exists in `cucim/python/cucim/src/cucim/_version.py` and when it is packaged, it would be rewritten & placed in `dist/cucim/_version.py` so only need to change `versionfile_build`

jakirkham replied:
Interesting. Ok let's leave it then :)

gigony commented on ci/gpu/build.sh:
```diff
@@ -77,13 +79,13 @@ gpuci_conda_retry build -c ${LIBCUCIM_BLD_PATH} -c conda-forge -c rapidsai-night
 # Install cuCIM and its dependencies
 gpuci_logger "Installing cuCIM and its dependencies"
-gpuci_conda_retry install -y -c ${LIBCUCIM_BLD_PATH} -c ${CUCIM_BLD_PATH} \
-    "rapids-build-env=$MINOR_VERSION.*" \
```
I found that there is no rapids-build-env package whose version is `$MINOR_VERSION.*` as we switched to CalVer.

https://anaconda.org/rapidsai/rapids-build-env/files

ajschmidt8 replied:
@gigony, this will need to be re-added. We use the `rapidsai-nightly` conda channel for pre-release packaging. You can find the cal-ver packages here: https://anaconda.org/rapidsai-nightly/rapids-build-env/files

jakirkham replied:
Somewhat related. I think we are missing an `activate` line (this is probably my fault). Was trying to fix that here ( https://github.com/rapidsai/cucim/pull/45/commits/495aa7316661b2efc95957d39c49f8831fb73f91 ). Think we may need to add that as well

gigony commented on cpp/include/cucim/cuimage.h:
```diff
@@ -17,19 +17,22 @@
-#include <set>
-
 #include "cucim/core/framework.h"
+
```
Thanks! Will remove it.

ajschmidt8 commented on CHANGELOG.md:
```diff
@@ -1,4 +1,4 @@
-# cuCIM 21.06.00 (Date TBD)
+# cuCIM [21.06.00](https://github.com/rapidsai/cucim/wiki/release_notes_v21.06.00) (Date TBD)
```
For consistency with the changelogs of other RAPIDS libraries, this change should probably be removed. Also, we have a script that updates the changelog during each release, so it would overwrite this change.

gigony replied:
Oh, I see. Will remove it from here.

ajschmidt8 reviewed (changes requested):
Few changes requested here. Also, `branch-21.06` is in code-freeze, so the base branch for this PR will need to be updated to `branch-21.08`.

quasiben commented:
> Few changes requested here. Also, branch-21.06 is in code-freeze, so the base branch for this PR will need to be updated to branch-21.08.

For clarity, we'd like for this to be released with 21.06 as other libraries would like to pick this up -- but not part of the larger RAPIDS integration meta-package.  Is that something we could do ?

ajschmidt8 commented:
> > Few changes requested here. Also, branch-21.06 is in code-freeze, so the base branch for this PR will need to be updated to branch-21.08.
> 
> For clarity, we'd like for this to be released with 21.06 as other libraries would like to pick this up -- but not part of the larger RAPIDS integration meta-package. Is that something we could do ?

@quasiben, gotcha. This PR has a _lot_ of changes, so I'm not sure it'd be a good idea to merge all of this a few days before the release. Couple questions here though.

1. Which libraries require these changes?
2. Are the changes here considered "required" for `21.06` to work? Or are they just "nice-to-haves"?

jakirkham commented:
Several of these changes were in previous PRs that were blocked due to CI issues. So Gigon has aggregated them together. IOW they are not as new as they might seem

The changes are really limited to cuCIM and are mainly in response to needs Clara has in their downstream libraries or outside user requests. So they shouldn't effect other RAPIDS libraries

quasiben commented:
@gigony would be better positioned to answer these questions:

1. Which libraries require these changes?
a. MONAI is a higher lever product depending on 
2. Are the changes here considered "required" for 21.06 to work? Or are they just "nice-to-haves"?
a. The majority of these changes are in the required for 21.06.  I would hope these large code changes before releases decreases as the library matures. 

Again, @gigony can you also comment here ?

gigony commented:
1.

MONAI and Clara Train SDK is using cuCIM.
since they are using PyPI package and I would release new version of PyPI
package(21.06.00) separately tomorrow, conda package doesn't affect to
those projects. However, not merging thid would cause discrepancy between
the two packages (conda vs PyPI) and that's what I am worrying about.

2.

I am sorry for the last minute changes. I will try to push all the needed
change before code freeze next time.



On Mon, Jun 7, 2021 at 11:52 AM Benjamin Zaitlen ***@***.***>
wrote:

> @gigony <https://github.com/gigony> would be better positioned to answer
> these questions:
>
>    1. Which libraries require these changes?
>    a. MONAI is a higher lever product depending on
>    2. Are the changes here considered "required" for 21.06 to work? Or
>    are they just "nice-to-haves"?
>    a. The majority of these changes are in the required for 21.06. I
>    would hope these large code changes before releases decreases as the
>    library matures.
>
> Again, @gigony <https://github.com/gigony> can you also comment here ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rapidsai/cucim/pull/48#issuecomment-856178008>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOW2SW4KQCVIOIWEAUI4PDTRUIOVANCNFSM46ENHSCA>
> .
>

raydouglass commented:
I think this is fine to merge this into `21.06` after addressing some of the request changed.

quasiben commented:
Thanks @raydouglass

gigony commented on cpp/include/cucim/cuimage.h:
```diff
@@ -162,12 +172,13 @@ class EXPORT_VISIBLE CuImage : public std::enable_shared_from_this<CuImage>
-    bool crop_image(io::format::ImageMetadataDesc* metadata,
-                    io::format::ImageReaderRegionRequestDesc* request,
-                    io::format::ImageDataDesc* out_image_data) const;
+    bool crop_image(io::format::ImageReaderRegionRequestDesc* request, io::format::ImageDataDesc* out_image_data) const;
```
Good points! Thanks. It makes sense. Updated it.

gigony commented on cpp/include/cucim/filesystem/file_handle.h:
```diff
@@ -37,11 +37,27 @@ enum class FileHandleType: uint16_t
+    EXPORT_VISIBLE CuCIMFileHandle(int fd,
+                                   CUfileHandle_t cufile,
+                                   FileHandleType type,
+                                   char* path,
```
`CuCIMFileHandle` is a C struct and used for C interface (https://github.com/gigony/cucim/blob/465c3ef4eef8367dd047c8a9492fed2076e0d717/cpp/include/cucim/io/format/image_format.h#L185) so wouldn't want to introduce C++ type there.
And currently value for `path` is dynamically allocated so cannot be `const` (https://github.com/gigony/cucim/blob/465c3ef4eef8367dd047c8a9492fed2076e0d717/cpp/plugins/cucim.kit.cuslide/src/cuslide/tiff/tiff.cpp#L247).


Will refactor pointer-related code as part of https://github.com/rapidsai/cucim/issues/38 in the future.

gigony commented on cpp/plugins/cucim.kit.cuslide/benchmarks/main.cpp:
```diff
@@ -86,7 +86,6 @@ static void test_basic(benchmark::State& state)
         auto handle = image_format->formats[0].image_parser.open(g_config.input_file.c_str());
 
         cucim::io::format::ImageMetadata metadata{};
-        metadata.level_count(1).level_downsamples({ 1.0 }).level_ndim(3);
```
Data in `metadata` would be overwritten by the following method (`image_format->formats[0].image_parser.parse(&handle, &metadata.desc());`) so no need to initialize.

```c
bool(CUCIM_ABI* parse)(CuCIMFileHandle* handle, ImageMetadataDesc* out_metadata);
```

```c++
        cucim::io::format::ImageMetadata metadata{};
        image_format->formats[0].image_parser.parse(&handle, &metadata.desc());
```
https://github.com/gigony/cucim/blob/465c3ef4eef8367dd047c8a9492fed2076e0d717/cpp/include/cucim/io/format/image_format.h#L193

gigony commented on cpp/plugins/cucim.kit.cuslide/src/cuslide/tiff/ifd.cpp:
```diff
@@ -446,25 +488,72 @@ bool IFD::read_region_tiles(const TIFF* tiff,
-                                                     tile_raster_nbytes, out_device);
-                }
+                // clang-format off
+                // cucim::logger::Timer tt("decode_deflate: {}\n");
+                //fmt::print(stderr, "# {}: {} {} - before key\n", std::chrono::high_resolution_clock::now().time_since_epoch().count(), getpid(), index); // [print]
```
Yes, no need for now, and can be refactored out for as benchmarking code.
Removed them.

gigony commented on cpp/plugins/cucim.kit.cuslide/src/cuslide/tiff/tiff.h:
```diff
@@ -17,21 +17,21 @@
 #include <cucim/io/format/image_format.h>
+#include <cucim/macros/api_header.h>
+#include <cucim/memory/memory_manager.h>
+//#include <tiffio.h>
```
Yes, we can. Removed.

gigony commented on cpp/src/cache/cache_type.cpp:
```diff
@@ -0,0 +1,101 @@
+// https://en.cppreference.com/w/cpp/algorithm/find
+#if __cplusplus < 202002L
+template <class InputIt, class UnaryPredicate>
+constexpr InputIt myfind_if(InputIt first, InputIt last, UnaryPredicate p)
```
Yes, constexpr is needed to implement constexpr map (Inspired by https://www.youtube.com/watch?v=INn3xa4pMfg).

The method is called at both compile time and runtime.

gigony commented on cpp/src/cache/image_cache_config.cpp:
```diff
@@ -0,0 +1,63 @@
+namespace cucim::cache
+{
+
+void ImageCacheConfig::load_config(void* json_obj)
```
Made it to `const`.
However, doesn't want to declare it with `json&` because we don't want to expose `json` type to the public header (https://github.com/gigony/cucim/blob/566f130425d2c4783788776c832d15783b4f665d/cpp/include/cucim/cache/image_cache_config.h#L43) which requires either cucim library package `nlohmann::json` header or application code depends on it.

gigony commented:
Thank you @quasiben @jakirkham @ajschmidt8 for the review!

I have addressed your comments except https://github.com/rapidsai/cucim/pull/48#discussion_r646838902 to make sure that this change is buildable without errors.

Could we merge this PR first and then apply @jakirkham 's patches shown below, to use rapids conda environment?

- https://github.com/rapidsai/cucim/commit/495aa7316661b2efc95957d39c49f8831fb73f91
- https://github.com/rapidsai/cucim/pull/45/files#diff-5ed0e3039eb400d13dc55579a034925ea8d08e9f2054085b1c79fdc46a7c09a8L29
